wire/msgtx.go

```diff
@@ -323,21 +448,45 @@ func (msg *MsgTx) Deserialize(r io.Reader) error {
 	// At the current time, there is no difference between the wire encoding
 	// at protocol version 0 and the stable long-term storage format.  As
 	// a result, make use of BtcDecode.
-	return msg.BtcDecode(r, 0)
+	return msg.BtcDecode(r, 0, WitnessEncoding)
+}
+
+// DeserializeNoWitness decodes a transaction from r into the receiver, where
+// the transaction encoding format within r MUST NOT utilize the new serialization
+// format created to encode transaction bearing witness data within inputs.
+func (msg *MsgTx) DeserializeNoWitness(r io.Reader) error {
+	return msg.BtcDecode(r, 0, BaseEncoding)
 }
 
 // BtcEncode encodes the receiver to w using the bitcoin protocol encoding.
 // This is part of the Message interface implementation.
 // See Serialize for encoding transactions to be stored to disk, such as in a
 // database, as opposed to encoding transactions for the wire.
-func (msg *MsgTx) BtcEncode(w io.Writer, pver uint32) error {
+func (msg *MsgTx) BtcEncode(w io.Writer, pver uint32, enc MessageEncoding) error {
 	var buf [4]byte
 	binary.LittleEndian.PutUint32(buf[:], uint32(msg.Version))
 	_, err := w.Write(buf[:])
 	if err != nil {
 		return err
 	}
 
+	// If the encoding version is set to WitnessEncoding, and the Flags
+	// field for the MsgTx aren't 0x00, then this indicates the transaction
+	// is to be encoded using the new witness inclusionary structure defined
+	// in BIP0141.
+	if enc == WitnessEncoding && msg.HasWitness() {
+		// After the txn's Version field, we include two additional
+		// bytes specific to the witness encoding. The first byte is an
+		// always 0x00 marker byte, which allows decoders to distinguish
+		// a serialized transaction with witnesses from a regular (legacy)
+		// one. The second byte is the Flag field, which at the moment is
+		// always 0x01, but way be extended in the future to accommodate
```

Comment:
// always 0x01, but ~~way~~ may be extended in the future to accommodate
// auxiliary non-commited fields.